Rust: add partial extractor for Expr, Statement, Pattern

aibaars opened the pull request:
Implement extraction of Expressions, Statements and Patterns.

redsun82 commented on rust/extractor/src/main.rs:
```diff
@@ -36,10 +51,16 @@ fn main() -> anyhow::Result<()> {
-        _macro_server.expect("no macro server");
+    let projects = find_project_manifests(&cfg.inputs).context("loading inputs")?;
+    for project in projects {
+        let mut workspace = ProjectWorkspace::load(project, &config, &progress)?;
+
+        if load_config.load_out_dirs_from_check {
+            let build_scripts = workspace.run_build_scripts(&config, &progress)?;
+            workspace.set_build_scripts(build_scripts)
+        }
+
+        let (db, vfs, _macro_server) = load_workspace(workspace, &config.extra_env, &load_config)?;
```
it's slightly weird to do an `if` on something hard-coded to `true `just 8 lines above this one. I understand you took this from `load_workspace_at` though. And, it does make it more robust if we ever tweak the configurations. Though to avoid the duplication from rust-analyzer code, maybe we can just call `load_workspace_at` on the manifests?
```suggestion
        let (db, vfs, _macro_server) = load_workspace_at(project, &config.extra_env, &load_config)?;
```
This will make a bit of unnecessary churning (rediscover the manifest itself), but it means we don't need to replicate `load_workspace_at` functionality in our code.

aibaars replied:
Good point, thanks!

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -64,6 +69,77 @@ impl CrateTranslator<'_> {
+                let start = data.line_index.line_col(range.start());
+                let end = data.line_index.line_col(range.end());
+                Some(self.trap.emit_location(data.label, start, end))
+            })
```
I _think_ the `source` can be taken by capture, without passing it around in the maps like this? Also, `clippy` told me already that `.and_then(... Some(x))` can really be just `.map(... x)`
```suggestion
        source
            .file_id
            .file_id()
            .map(|f| f.file_id())
            .and_then(|file_id| self.emit_file(file_id))
            .map(|data| {
                let range = source.value.text_range();
                let start = data.line_index.line_col(range.start());
                let end = data.line_index.line_col(range.end());
                self.trap.emit_location(data.label, start, end)
            })
```

aibaars replied:
Indeed, I think I already did that in a later commit.

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
             })
     }
 
+    fn emit_label(
```
uh, I can see we could get in nomenclature confusion between trap labels and rust labels... maybe we should `s/label/trap_label` where it applies to avoid that?

aibaars replied:
I'm not sure I understand . This function does not emit a trap label .

redsun82 replied:
Ah, sorry, my bad, I got confused: we actually don't have any function here named with `label` in it and referring to trap labels. All is good :slightly_smiling_face:

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -64,6 +69,77 @@ impl CrateTranslator<'_> {
         })
     }
 
+    fn emit_location_ast_ptr<T>(
```
minor bike-shedding naming nit: maybe here and in the following `emit_ast_ptr_location`, or `emit_location_for_ast_ptr`? Like it is it could be read "emit the ast_ptr for a location"

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                    .into_iter()
+                    .map(|pat| self.emit_pat(*pat, body, source_map))
+                    .collect();
+                let ellipsis = ellipsis.and_then(|x| x.try_into().ok());
```
I think it might help understanding naming this `ellipsis_index`

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                    id: TrapId::Star,
+                    location,
+                    expr,
+                    has_semi: *has_semi,
```
I'm guessing this means semicolon? If yes, it might be worth spelling it out entirely

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                    has_semi: *has_semi,
+                })
+            }
+            Statement::Item => self.trap.emit(generated::ItemStmt {
```
this one's weird (on the side of rust-analyzer, not yours :slightly_smiling_face:) maybe we don't really need it? We'll have to look into it.

aibaars replied:
I think it is a useless node. I just added it for completeness, so we can use QL to figure out when these nodes occur and decide if they are truly useless or not.

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                id: TrapId::Star,
+                location,
+            }),
+            ra_ap_hir_def::hir::Expr::Path(_) => self.trap.emit(generated::Path {
+                id: TrapId::Star,
+                location,
+            }),
```
in Swift we had an overcomplicated thing to track "yet unextracted" entities. I'm thinking if it's not worth to simply add a 
```python
class NotImplementedExpr(Expr):
    kind: str
```
to track that.

redsun82 replied:
The good thing about `NotImplementedExpr` or a similar solution in the DB, is that we can query and set up telemetry for it

redsun82 replied:
(but we can do that in a follow-up PR)

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                let condition = self.emit_expr(*condition, &body, &source_map);
+                let then = self.emit_expr(*then_branch, &body, &source_map);
+                let else_ = else_branch.map(|x| self.emit_expr(x, &body, &source_map));
+                self.trap.emit(generated::If {
```
Maybe we want to suffix all `Expr` on the QL/codegen side? I.e. `IfExpr`, `LetExpr`, etc? You did that with `Pat`, migth be worth doing it for `Expr` and `Stmt` as well.

aibaars replied:
Yes, that is probably most consistent. I had to suffix some of the others to avoid name clashes.

redsun82 reviewed (commented):
wow, great stuff! I have mostly nits and naming suggestions, but this looks good!

paldepind commented on rust/schema.py:
```diff
@@ -480,19 +480,19 @@ class InlineAsm(Expr):
-    type_ref: optional[TypeRef]
-    initializer: optional[Expr]
-    else_: optional[Expr]
+class LetStmt(Stmt):
```
Is this a [`let`](https://doc.rust-lang.org/std/keyword.let.html) or an [`if let`](https://doc.rust-lang.org/reference/expressions/if-expr.html#if-let-expressions)? The old name and the fact that it has an else branch makes it look like the latter. If that's true I think `IfLet` should still be in the name?

aibaars replied:
That's what I thought too, but I was wrong. That's why I changed the name. The use of `else` is very rare, but I found some examples using QL:

https://github.com/rust-lang/rust/blob/304b7f801bab31233680879ca4fb6eb294706a59/library/alloc/src/vec/in_place_collect.rs#L322

paldepind replied:
I see. Thanks for the explanation, I had no idea that a `let` could have and `else` branch. Looking it up, the docs state that "the `else` block must always diverge" so that's useful to be aware of for the CFG.

redsun82 replied:
> I had no idea that a `let` could have and `else` branch

oh, me neither!

aibaars commented:
@paldepind @redsun82 I addressed most comments. I added the `Expr` suffix to all expression classes, please make sure to update any QL code you may have.

paldepind commented:
> I added the `Expr` suffix to all expression classes, please make sure to update any QL code you may have.

Done 👌 and thanks. PR seems great from my perspective. I have played around with printing ASTs and it seems to me that there's enough in order to do CFGs.

redsun82 commented on rust/extractor/src/main.rs:
```diff
@@ -42,6 +42,9 @@ fn main() -> anyhow::Result<()> {
 
     let config = CargoConfig {
         sysroot: Some(RustLibSource::Discover),
+        target_dir: std::env::var("CODEQL_EXTRACTOR_RUST_SCRATCH_DIR")
+            .map(|p| ra_ap_paths::Utf8PathBuf::from(p).join("target"))
+            .ok(),
```
I would rather add it to `config.rs`, like the other directories, so that we can also pass `--scratch_dir` to tweak this.

Also, my memory is hazy, but I think that despite the extractor-specific env variable name, the scratch dir is actually shared when building in cluster mode, so we might need to use `rust/target` for robustness, but I'll try to double check.

redsun82 commented on rust/codeql-extractor.yml:
```diff
@@ -13,4 +13,24 @@ file_types:
+        title: Controls compression for the TRAP files written by the extractor.
+        description: >
+          This option is only intended for use in debugging the extractor. Accepted
+          values are 'brotli' (the default, to write brotli-compressed TRAP), 'gzip', and 'none'
```
uh, do we really support brotli?

aibaars replied:
Probably not, I guess I should have copied that snippet from ruby instead of c# ;-)

redsun82 replied:
food for thought: we could add `zstd` support, that can speed up the trap import phase (zstd is really fast at decompression).

redsun82 commented on rust/extractor/src/config.rs:
```diff
@@ -72,6 +73,7 @@ impl Config {
         }
         Figment::new()
             .merge(Env::prefixed("CODEQL_EXTRACTOR_RUST_"))
+            .merge(Env::prefixed("CODEQL_EXTRACTOR_RUST_OPTION_"))
```
I like it!

redsun82 commented on rust/extractor/src/translate.rs:
```diff
@@ -928,20 +931,27 @@ impl CrateTranslator<'_> {
+                let body = if self.extract_dependencies || self.krate.origin(self.db).is_local() {
+                    let (body, source_map) = self.db.body_with_source_map(def.into());
+                    let txt = body.pretty_print(self.db, def.into(), Edition::Edition2021);
+                    println!("{}", &txt);
```
this is a debug print I guess?

We might want to sprinkle some `trace!` logs maybe at some point (not in scope now I think)

aibaars replied:
Yes, this should be removed at some point. I found it quite useful to see the expanded HIR code.

redsun82 replied:
in a previous iteration of the code that was not using the `trap` library, I was `trace!`ing everything emitted. Might be worth replacing the `println!` with that and keeping it in. There should be a way to get verbose mode out with env variables

redsun82 replied:
(but then again, this can be done as a follow-up, we can keep the `println!` now)

redsun82 replied:
> There should be a way to get verbose mode out with env variables

yeah, just `CODEQL_EXTRACTOR_RUST_VERBOSE=2`

aibaars commented on rust/extractor/src/config.rs:
```diff
@@ -32,6 +32,7 @@ pub struct Config {
     pub scratch_dir: PathBuf,
     pub trap_dir: PathBuf,
     pub source_archive_dir: PathBuf,
+    pub extract_dependencies: bool,
```
It's already a CodeQL extractor option which can be passed to `codeql database create` . Not sure if there is a point having all those flags repeated in the extractor binary too.

redsun82 replied:
from experience it can be nice to try out the extractor in isolation from `codeql`, and without writing all that screaming snake case in the env :D. Also, `qltest.sh` allows to inject extractor flags in the test sources via comments, so it can be useful to have CLI flags. But we can also inject env variables there, so this is not relevant.
